sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamRealTimeModeAllowlistSuite.scala

```diff
@@ -0,0 +1,163 @@
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.internal.SQLConf
+
+class StreamRealTimeModeAllowlistSuite extends StreamRealTimeModeE2ESuiteBase {
+  import testImplicits._
+
+  test("rtm source allowlist") {
+    val query = spark.readStream
+      .format("rate")
+      .option("numPartitions", 1)
+      .load()
+      .writeStream
+      .format("console")
+      .outputMode("update")
+      .trigger(defaultTrigger)
+      .start()
+
+    eventually(timeout(60.seconds)) {
+      checkError(
+        exception = query.exception.get.getCause.asInstanceOf[SparkIllegalArgumentException],
+        condition = "STREAMING_REAL_TIME_MODE.INPUT_STREAM_NOT_SUPPORTED",
+        parameters = Map(
+          "className" ->
+            "org.apache.spark.sql.execution.streaming.sources.RateStreamMicroBatchStream")
+      )
+    }
+  }
+
+  test("rtm operator allowlist") {
+    withSQLConf("spark.sql.autoBroadcastJoinThreshold" -> "0") {
+      val inputData = LowLatencyMemoryStream[Int](2)
+      val staticDf = spark.range(1, 31, 1, 10).selectExpr("id AS join_key", "id AS join_value")
+
+      val df = inputData.toDF()
+        .select(col("value").as("key"), col("value").as("value"))
+        .join(staticDf, col("value") === col("join_key"))
+        .select(
+          concat(col("key"), lit("-"), col("value"), lit("-"), col("join_value")).as("output"))
+
+      val query = runStreamingQuery("operation_allowlist", df)
+
+      eventually(timeout(60.seconds)) {
+        checkError(
+          exception = query.exception.get.getCause.asInstanceOf[SparkIllegalArgumentException],
+          condition = "STREAMING_REAL_TIME_MODE.OPERATOR_OR_SINK_NOT_IN_ALLOWLIST",
+          parameters = Map(
+            "errorType" -> "operator",
+            "message" -> (
+              "org.apache.spark.sql.execution.SortExec, " +
+                "org.apache.spark.sql.execution.exchange.ShuffleExchangeExec, " +
+                "org.apache.spark.sql.execution.joins.SortMergeJoinExec are"
+              )
+          )
+        )
+      }
+    }
+  }
+
+  test("rtm sink allowlist") {
+    val read = LowLatencyMemoryStream[Int](2)
+
+    val query = read
+      .toDF()
+      .writeStream
+      .format("noop")
+      .outputMode(OutputMode.Update())
+      .trigger(defaultTrigger)
+      .queryName("rtm_sink_allowlist")
+
+    checkError(
+      exception = intercept[SparkIllegalArgumentException] {
+        query.start()
+      },
+      condition = "STREAMING_REAL_TIME_MODE.OPERATOR_OR_SINK_NOT_IN_ALLOWLIST",
+      parameters = Map(
+        "errorType" -> "sink",
+        "message" -> "org.apache.spark.sql.execution.datasources.noop.NoopTable$ is"
+      ))
+
+    withSQLConf(SQLConf.STREAMING_REAL_TIME_MODE_ALLOWLIST_CHECK.key -> "false") {
+      val tmp = query.start()
+      Thread.sleep(5000)
+      tmp.stop()
+    }
+  }
+
+  // TODO : Remove this test after RTM can shuffle to multiple stages
```

Comment:
Please file a JIRA issue and make this IDed TODO like `TODO(SPARK-XXX)`. It helps the community move forward from your contributions.